src/customloaders.cpp

```diff
@@ -0,0 +1,96 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "customloaders.hpp"
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <cxxopts.hpp>
+#include <spdlog/spdlog.h>
+
+#include "customloaderinterface.hpp"
+
+namespace ovms {
+
+Status CustomLoaders::add(std::string name, std::shared_ptr<CustomLoaderInterface> loaderInsterface, void* library) {
+    SPDLOG_INFO("Adding loder {} to loaders list", name);
+    auto loaderIt = newCustomLoaderInterfacePtrs.find(name);
+    if (loaderIt == newCustomLoaderInterfacePtrs.end()) {
+        newCustomLoaderInterfacePtrs.insert({name, std::make_pair(library, loaderInsterface)});
+        return StatusCode::OK;
+    }
+
+    return StatusCode::CUSTOM_LOADER_EXISTS;
+}
+
+Status CustomLoaders::remove(const std::string& name) {
+    SPDLOG_INFO("Removing loder {} from loaders list", name);
+    auto loaderIt = customLoaderInterfacePtrs.find(name);
+
+    if (loaderIt == customLoaderInterfacePtrs.end()) {
+        return StatusCode::CUSTOM_LOADER_NOT_PRESENT;
+    }
+
+    customLoaderInterfacePtrs.erase(loaderIt);
+    return StatusCode::OK;
+}
+
+std::shared_ptr<CustomLoaderInterface> CustomLoaders::find(const std::string& name) {
+    SPDLOG_INFO("looking for loder {} in loaders list", name);
+    auto loaderIt = customLoaderInterfacePtrs.find(name);
+
+    if (loaderIt == customLoaderInterfacePtrs.end()) {
+        return nullptr;
+    }
+
+    return (loaderIt->second).second;
+}
+
+Status CustomLoaders::move(const std::string& name) {
+    SPDLOG_INFO("Moving loder {} from old to new loaders list", name);
+    auto loaderIt = customLoaderInterfacePtrs.find(name);
+
+    if (loaderIt == customLoaderInterfacePtrs.end()) {
+        return StatusCode::INTERNAL_ERROR;
+    }
+
+    newCustomLoaderInterfacePtrs.insert({name, customLoaderInterfacePtrs[name]});
+    customLoaderInterfacePtrs.erase(loaderIt);
+    return StatusCode::OK;
+}
+
+Status CustomLoaders::finalize() {
+    // By now the remaining loaders in current list are not there in new config. Delete them
+    for (auto it = customLoaderInterfacePtrs.begin(); it != customLoaderInterfacePtrs.end(); it++) {
+        SPDLOG_INFO("Loader {} is not there in new list.. deleting the same", it->first);
+        auto loaderPtr = (it->second).second;
+        loaderPtr->loaderDeInit();
+    }
+
+    SPDLOG_INFO("Clearing the list");
+    customLoaderInterfacePtrs.clear();
+
+    SPDLOG_INFO("Adding new list to  the old list");
```

Comment:
Double space

Reply:
Additionally - could we give more information here to the user which custom loader is doing action here? This would ease debugging.

Reply:
Done